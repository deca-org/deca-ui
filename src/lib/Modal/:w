import { CSS } from '@lib/Theme';
import { useDOMRef, mergeRefs, useClickOutside } from '@lib/Utils';
import { useTransition } from '@react-spring/web';
import clsx from 'clsx';
import React, { useEffect, SetStateAction, Dispatch } from 'react';
import ReactDOM from 'react-dom';

import { StyledModal, StyledModalOverlay } from './Modal.styles';

export interface ModalProps {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /*
   * If true, the component is shown.
   */
  open?: boolean;
  /**
   * State dispatcher function (setter in useState)
   */
  setOpen?: Dispatch<SetStateAction<boolean>>;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
}

const Modal = React.forwardRef(
  (
    { children, open, setOpen, css, className = '' }: ModalProps,
    ref: React.Ref<HTMLDivElement | null>
  ) => {
    const handleEsc = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        setOpen && setOpen(false);
      }
    };

    const modalRef = useDOMRef(ref);

    useEffect(() => {
      window.addEventListener('keydown', handleEsc);
      return () => {
        window.removeEventListener('keydown', handleEsc);
      };
    }, []);

    const transition = useTransition(open, {
      from: {
        scale: 0.75,
        opacity: 0,
        position: 'fixed',
        top: '55%',
        left: '50%',
        transform: 'translate(-50%, -45%)',
      },
      enter: {
        scale: 1,
        opacity: 1,
        position: 'fixed',
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
      },
      leave: {
        scale: 0.75,
        opacity: 0,
        position: 'fixed',
        top: '55%',
        left: '50%',
        transform: 'translate(-50%, -45%)',
      },
      config: {
        tension: 300,
        friction: 15,
      },
    });

    const overlayTransition = useTransition(open, {
      from: {
        opacity: 0,
      },
      enter: {
        scale: 1,
        opacity: 1,
      },
      leave: {
        opacity: 0,
      },
      config: {
        duration: 120,
      },
    });

    const clickOutsideRef = useClickOutside(() => {
      setOpen && setOpen(false);
    });

    const preClass = 'decaModal';

    return ReactDOM.createPortal(
      transition(
        (style, item) =>
          item && (
            <>
              {overlayTransition(
                (overlayStyle, overlayItem) =>
                  overlayItem && <StyledModalOverlay style={overlayStyle} />
              )}
              <StyledModal
                ref={mergeRefs(modalRef, clickOutsideRef)}
                style={style}
                className={clsx(className, `${preClass}-root`)}
                css={css}
              >
                {children}
              </StyledModal>
            </>
          )
      ),
      document.querySelector('body') as Element
    );
  }
);

export default Modal;
