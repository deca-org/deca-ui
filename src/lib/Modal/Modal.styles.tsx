import { styled, theme } from '@lib/Theme/stitches.config';
import { animated } from '@react-spring/web';
import { transparentize } from 'polished';

export const StyledModalOverlay = styled(animated.div, {
  position: 'fixed',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  bg: transparentize(0.4, theme.colors.black.value),
  zIndex: '$10',
});

export const StyledModal = styled(animated.div, {
  position: 'fixed',
  fontFamily: '$normal',
  boxShadow: '$default',
  br: '$sm',
  zIndex: '$max',
  color: '$text',
  variants: {
    noPadding: {
      true: {
        p: '$n',
      },
      false: {
        p: '$3',
      },
    },
    isDark: {
      true: {
        bg: '$popperDarkBg',
      },
      false: {
        bg: '$popperLightBg',
      },
    },
  },
});

export const StyledModalFlexbox = styled('div', {
  display: 'flex',
  justifyContent: 'center',
  flexDirection: 'column',
  variants: {
    autoGap: {
      true: {
        gap: '$4',
      },
      false: {
        gap: '$n',
      },
    },
  },
});

export const StyledModalHeader = styled('div', {
  color: '$text',
  display: 'flex',
  justifyContent: 'center',
  flexDirection: 'column',
  variants: {
    autoGap: {
      true: {
        gap: '$2',
      },
      false: {
        gap: '$n',
      },
    },
  },
});

export const StyledModalBody = styled('div', {
  color: '$text',
  display: 'flex',
  justifyContent: 'center',
  flexDirection: 'column',
  variants: {
    autoGap: {
      true: {
        gap: '$2',
      },
      false: {
        gap: '$n',
      },
    },
  },
});

export const StyledModalFooter = styled('div', {
  color: '$text',
  display: 'flex',
  justifyContent: 'flex-end',
  flexDirection: 'row',
  variants: {
    autoGap: {
      true: {
        gap: '$2',
      },
      false: {
        gap: '$n',
      },
    },
  },
});
