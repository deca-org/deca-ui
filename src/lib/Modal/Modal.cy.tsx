import { Test } from '../Utils';
import _cyp from '../../../cypress';
import React from 'react';
import Modal, {
  ModalProps,
  ModalHeaderProps,
  ModalBodyProps,
  ModalFooterProps,
} from '.';
import Button from '../Button';
import Text from '../Text';
import Input from '../Input';
import DecaUIProvider from '../Theme';

interface ModalComposerProps extends ModalProps {
  header?: ModalHeaderProps;
  body?: ModalBodyProps;
  footer?: ModalFooterProps;
}

const ModalComposer = (props: ModalComposerProps) => (
  <Modal closeButton {...props} data-testid="test.modal">
    <Modal.Header {...props.header}>
      <Text as="p" size="$h6" center>
        Welcome to <b>DecaUI</b>
      </Text>
    </Modal.Header>
    <Modal.Body {...props.body}>
      <Input variant="outlined" placeholder="Email Address" size="lg" />
      <Input variant="outlined" placeholder="Password" size="lg" />
    </Modal.Body>
    <Modal.Footer {...props.footer}>
      <Button variant="flat" color="error">
        Close
      </Button>
      <Button>Sign Up</Button>
    </Modal.Footer>
  </Modal>
);

const modalSelector = '[data-testid="test.modal"]';

describe('components/Modal', () => {
  describe('base', () => {
    it('background-color', () => {
      cy.mount(<ModalComposer open={true} />);
      cy.get(modalSelector).should(
        'have.css',
        'background-color',
        Test.color('white')
      );
    });
  });
  describe('state', () => {
    it('not exist in DOM when not open', () => {
      cy.mount(<ModalComposer open={false} />);
      cy.get(modalSelector).should('not.exist');
    });
    it('exists in DOM when open', () => {
      cy.mount(<ModalComposer open={true} />);
      cy.get(modalSelector).should('exist');
    });
  });
  describe('autoGap', () => {
    it('container', () => {
      cy.mount(<ModalComposer open={true} />);
      cy.get('.decaModal-flexbox').should('have.css', 'gap', Test.space('4'));
    });
    it('header', () => {
      cy.mount(<ModalComposer open={true} />);
      cy.get('.decaModalHeader-root').should(
        'have.css',
        'gap',
        Test.space('2')
      );
    });
    it('body', () => {
      cy.mount(<ModalComposer open={true} />);
      cy.get('.decaModalBody-root').should('have.css', 'gap', Test.space('2'));
    });
    it('footer', () => {
      cy.mount(<ModalComposer open={true} />);
      cy.get('.decaModalFooter-root').should(
        'have.css',
        'gap',
        Test.space('2')
      );
    });
  });
  describe('disabled autoGap', () => {
    it('container', () => {
      cy.mount(<ModalComposer open={true} autoGap={false} />);
      cy.get('.decaModal-flexbox').should('have.css', 'gap', Test.space('n'));
    });
    it('header', () => {
      cy.mount(<ModalComposer open={true} header={{ autoGap: false }} />);
      cy.get('.decaModalHeader-root').should(
        'have.css',
        'gap',
        Test.space('n')
      );
    });
    it('body', () => {
      cy.mount(<ModalComposer open={true} body={{ autoGap: false }} />);
      cy.get('.decaModalBody-root').should('have.css', 'gap', Test.space('n'));
    });
    it('footer', () => {
      cy.mount(<ModalComposer open={true} footer={{ autoGap: false }} />);
      cy.get('.decaModalFooter-root').should(
        'have.css',
        'gap',
        Test.space('n')
      );
    });
  });

  describe('padding', () => {
    it('default padding', () => {
      cy.mount(<ModalComposer open={true} />);
      cy.get(modalSelector).should('have.css', 'padding', Test.space('3'));
    });
    it('noPadding', () => {
      cy.mount(<ModalComposer open={true} noPadding />);
      cy.get(modalSelector).should('have.css', 'padding', Test.space('n'));
    });
  });
});
