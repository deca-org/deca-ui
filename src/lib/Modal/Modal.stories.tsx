import Button from '@lib/Button';
import Input from '@lib/Input';
import Modal from '@lib/Modal';
import Text from '@lib/Text';
import { DecaUIProvider } from '@lib/Theme';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React, { useState } from 'react';

export default {
  title: 'Modal',
  component: Modal,
} as ComponentMeta<typeof Modal>;

const Template: ComponentStory<typeof Modal> = (args) => {
  const [open, setOpen] = useState(false);

  return (
    <>
      <Button onClick={() => setOpen(true)}>Open Modal</Button>
      <Modal closeButton open={open} setOpen={setOpen} {...args}>
        <Modal.Header>
          <Text as="p" size="h6" center>
            Welcome to <b>DecaUI</b>
          </Text>
        </Modal.Header>
        <Modal.Body>
          <Input variant="outlined" placeholder="Email Address" size="lg" />
          <Input variant="outlined" placeholder="Password" size="lg" />
        </Modal.Body>
        <Modal.Footer>
          <Button variant="flat" color="error" onClick={() => setOpen(false)}>
            Close
          </Button>
          <Button>Sign Up</Button>
        </Modal.Footer>
      </Modal>
    </>
  );
};

export const Default = Template.bind({});
Default.args = {
  noPadding: false,
  autoGap: true,
  closeButton: true,
};

export const WithTheme = Template.bind({});
WithTheme.args = { ...Default.args };
WithTheme.decorators = [
  (Story) => (
    <DecaUIProvider
      theme={{
        colors: {
          primary: '$green600',
        },
        radii: {
          sm: '15px',
        },
      }}
    >
      <Story />
    </DecaUIProvider>
  ),
];

export const DarkMode = Template.bind({});

DarkMode.args = { ...Default.args };
DarkMode.parameters = { backgrounds: { default: 'dark' } };

DarkMode.decorators = [
  (Story) => (
    <DecaUIProvider mode="dark">
      <Story />
    </DecaUIProvider>
  ),
];
