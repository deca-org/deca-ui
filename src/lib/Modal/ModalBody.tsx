import { CSS } from '@lib/Theme';
import {
  __DEV__,
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
} from '@lib/Utils';
import clsx from 'clsx';
import React, { useContext } from 'react';

import { ModalContext, IModalContext } from './Modal';
import { StyledModalBody } from './Modal.styles';

/**
 * ModalBody contains the main content of a modal component
 */
interface Props {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * Have gap between all elements.
   */
  autoGap?: boolean;
}

export type ModalBodyProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type ModalBodyComponent = (<C extends React.ElementType = 'div'>(
  props: ModalBodyProps<C>
) => React.ReactElement | null) & { displayName?: string };

const ModalBody: ModalBodyComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    { children, className = '', css, as, autoGap, ...props }: ModalBodyProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const context = useContext(ModalContext) as IModalContext;

    const preClass = 'decaModalBody';

    return (
      <StyledModalBody
        autoGap={autoGap !== undefined ? autoGap : context.autoGap}
        ref={ref}
        className={clsx(className, `${preClass}-root`)}
        as={as}
        css={css}
        {...props}
      >
        {children}
      </StyledModalBody>
    );
  }
);

if (__DEV__) {
  ModalBody.displayName = 'DecaUI.ModalBody';
}

export default ModalBody;
