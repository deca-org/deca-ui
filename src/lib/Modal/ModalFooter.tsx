import { CSS } from '@lib/Theme/stitches.config';
import {
  __DEV__,
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
} from '@lib/Utils';
import clsx from 'clsx';
import React, { useContext } from 'react';

import { ModalContext, IModalContext } from './Modal';
import { StyledModalFooter } from './Modal.styles';

/**
 * ModalFooter allows users to place content on the bottom of their modal component
 */
interface Props {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * Have gap between all elements.
   */
  autoGap?: boolean;
}

export type ModalFooterProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type ModalFooterComponent = (<C extends React.ElementType = 'div'>(
  props: ModalFooterProps<C>
) => React.ReactElement | null) & { displayName?: string };

const ModalFooter: ModalFooterComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    {
      children,
      className = '',
      css,
      as,
      autoGap,
      ...props
    }: ModalFooterProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const context = useContext(ModalContext) as IModalContext;

    const preClass = 'decaModalFooter';

    return (
      <StyledModalFooter
        autoGap={autoGap !== undefined ? autoGap : context.autoGap}
        ref={ref}
        className={clsx(className, `${preClass}-root`)}
        as={as}
        css={css}
        {...props}
      >
        {children}
      </StyledModalFooter>
    );
  }
);

if (__DEV__) {
  ModalFooter.displayName = 'DecaUI.ModalFooter';
}

export default ModalFooter;
