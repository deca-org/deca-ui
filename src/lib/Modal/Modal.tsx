import { ThemeContext } from '@lib/Theme';
import Box from '@lib/Box';
import Button from '@lib/Button';
import { CSS } from '@lib/Theme/stitches.config';
import {
  mergeRefs,
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
  useClickOutside,
  __DEV__,
  MasterComponent,
} from '@lib/Utils';
import { animated, useTransition } from '@react-spring/web';
import clsx from 'clsx';
import React, { useEffect, useState, SetStateAction, Dispatch } from 'react';
import ReactDOM from 'react-dom';

import {
  StyledModal,
  StyledModalFlexbox,
  StyledModalOverlay,
} from './Modal.styles';

import ModalHeader from './ModalHeader';
import ModalBody from './ModalBody';
import ModalFooter from './ModalFooter';

/*
 * The Modal component provides a foundation for creating dialogs or popovers.
 */
interface Props {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /*
   * If true, the component is shown.
   */
  open?: boolean;
  /**
   * State dispatcher function (setter in useState)
   */
  setOpen?: Dispatch<SetStateAction<boolean>>;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * Have close button appear at the top right corner of the modal.
   */
  closeButton?: boolean;
  /**
   * Have gap between all elements.
   * @default true
   */
  autoGap?: boolean;
  /**
   * Remove padding applied to Modal component.
   * @default false
   */
  noPadding?: boolean;
}

export interface IModalContext {
  setOpen?: Dispatch<SetStateAction<boolean>>;
  autoGap?: boolean;
}

export const ModalContext = React.createContext<IModalContext | null>(null);

const CloseButton = () => (
  <svg
    viewBox="0 0 24 24"
    aria-hidden="true"
    focusable="false"
    fill="currentColor"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g data-name="Layer 2">
      <path
        d="M13.41 12l4.3-4.29a1 1 0 10-1.42-1.42L12 10.59l-4.29-4.3a1 1 0 00-1.42 1.42l4.3 4.29-4.3 4.29a1 1 0 000 1.42 1 1 0 001.42 0l4.29-4.3 4.29 4.3a1 1 0 001.42 0 1 1 0 000-1.42z"
        data-name="close"
      ></path>
    </g>
  </svg>
);

export type ModalProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type ModalComponent = (<C extends React.ElementType = 'div'>(
  props: ModalProps<C>
) => React.ReactElement | null) & { displayName?: string };

const Modal: ModalComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    {
      children,
      open,
      setOpen,
      css,
      className = '',
      closeButton = false,
      autoGap = true,
      noPadding = false,
      as,
      ...props
    }: ModalProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const handleEsc = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        setOpen && setOpen(false);
      }
    };

    useEffect(() => {
      window.addEventListener('keydown', handleEsc);
      return () => {
        window.removeEventListener('keydown', handleEsc);
      };
    }, []);

    const transition = useTransition(open, {
      from: {
        scale: 0.75,
        opacity: 0,
      },
      enter: {
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
        scale: 1,
        opacity: 1,
      },
      leave: {
        scale: 0.75,
        opacity: 0,
      },
      config: {
        tension: 220,
        friction: 18,
      },
    });

    const overlayTransition = useTransition(open, {
      from: {
        opacity: 0,
      },
      enter: {
        scale: 1,
        opacity: 1,
      },
      leave: {
        opacity: 0,
      },
      config: {
        duration: 120,
      },
    });

    const clickOutsideRef = useClickOutside(() => {
      setOpen && setOpen(false);
    });

    const preClass = 'decaModal';

    const { dark } = React.useContext(ThemeContext);

    const [DOM, setDOM] = useState(false);

    useEffect(() => {
      setDOM(true);
    }, []);

    if (DOM) {
      return ReactDOM.createPortal(
        transition(
          (style, item) =>
            item && (
              <ModalContext.Provider value={{ setOpen, autoGap }}>
                {overlayTransition(
                  (overlayStyle, overlayItem) =>
                    overlayItem && <StyledModalOverlay style={overlayStyle} />
                )}
                <StyledModal
                  ref={mergeRefs(ref, clickOutsideRef)}
                  style={style}
                  className={clsx(className, `${preClass}-root`)}
                  css={css}
                  noPadding={noPadding}
                  as={animated[as as keyof JSX.IntrinsicElements]}
                  isDark={dark}
                  {...props}
                >
                  {closeButton && (
                    <Box
                      css={{
                        display: 'flex',
                        justifyContent: 'flex-end',
                      }}
                    >
                      <Button
                        variant="ghost"
                        pill
                        icon={<CloseButton />}
                        css={{
                          mr: noPadding ? '-$0' : '-$3',
                          mt: noPadding ? '-$0' : '-$3',
                          p: '$n',
                          bg: 'transparent',
                          color: dark ? '$gray700' : '$gray500',
                          '&:hover': {
                            color: '$gray600',
                            bg: 'transparent',
                            borderColor: '$transparent',
                          },
                          '&:focus': {
                            color: dark ? '$gray500' : '$gray700',
                            bg: 'transparent',
                          },
                        }}
                        onClick={() => setOpen && setOpen(false)}
                      />
                    </Box>
                  )}
                  <StyledModalFlexbox
                    className={`${preClass}-flexbox`}
                    autoGap={autoGap}
                  >
                    {children}
                  </StyledModalFlexbox>
                </StyledModal>
              </ModalContext.Provider>
            )
        ),
        document.getElementById('decaUI-provider')
          ? (document.getElementById('decaUI-provider') as Element)
          : (document.querySelector('body') as Element)
      );
    }
    return <></>;
  }
);

if (__DEV__) {
  Modal.displayName = 'DecaUI.Modal';
}

export default Modal as MasterComponent<
  HTMLDivElement,
  ModalProps<React.ElementType>,
  {
    Header: typeof ModalHeader;
    Body: typeof ModalBody;
    Footer: typeof ModalFooter;
  }
>;
