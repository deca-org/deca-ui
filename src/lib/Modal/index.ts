import Modal from './Modal';
import ModalBody from './ModalBody';
import ModalFooter from './ModalFooter';
import ModalHeader from './ModalHeader';

Modal.Header = ModalHeader;
Modal.Body = ModalBody;
Modal.Footer = ModalFooter;

export * from './Modal';
export { Modal };
export default Modal;
