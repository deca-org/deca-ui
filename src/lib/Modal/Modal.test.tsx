import Modal from '@lib/Modal';
import { render, screen } from '@testing-library/react';
import React from 'react';

describe('components/Modal', () => {
  global.ResizeObserver = require('resize-observer-polyfill');
  it('matches snapshot', () => {
    const { asFragment } = render(<Modal open>Hello World</Modal>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders text when modal is open', () => {
    render(<Modal open>content</Modal>);
    expect(screen.queryByText('content')).not.toBe(null);
  });
  it('does not render text when modal is closed', () => {
    render(<Modal>content</Modal>);
    expect(screen.queryByText('content')).toBe(null);
  });
  it('modal components all render correctly', () => {
    const { asFragment } = render(
      <Modal open>
        <Modal.Header>Header component</Modal.Header>
        <Modal.Body>Header component</Modal.Body>
        <Modal.Footer>Header component</Modal.Footer>
      </Modal>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
