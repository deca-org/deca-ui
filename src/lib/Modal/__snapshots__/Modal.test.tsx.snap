// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modal matches snapshot 1`] = `<DocumentFragment />`;

exports[`components/Modal modal components all render correctly 1`] = `<DocumentFragment />`;
